lib/pluginmanager/bundler/logstash_uninstall.rb

```diff
@@ -34,54 +34,50 @@ def initialize(gemfile_path, lockfile_path)
       @lockfile_path = lockfile_path
     end
 
-    # To be uninstalled the candidate gems need to be standalone.
-    def dependants_gems(gem_name)
-      builder = Dsl.new
-      builder.eval_gemfile(::File.join(::File.dirname(gemfile_path), "original gemfile"), File.read(gemfile_path))
-      definition = builder.to_definition(lockfile_path, {})
-
-      definition.specs
-        .select { |spec| spec.dependencies.collect(&:name).include?(gem_name) }
-        .collect(&:name).sort.uniq
-    end
-
-    def uninstall!(gem_name)
-      unfreeze_gemfile do
-        dependencies_from = dependants_gems(gem_name)
-
-        if dependencies_from.size > 0
-          display_cant_remove_message(gem_name, dependencies_from)
-          false
-        else
-          remove_gem(gem_name)
-          true
+    def uninstall!(gems_to_remove)
+      gems_to_remove = Array(gems_to_remove)
+
+      unsatisfied_dependency_mapping = Dsl.evaluate(gemfile_path, lockfile_path, {}).specs.each_with_object({}) do |spec, memo|
+        next if gems_to_remove.include?(spec.name)
+        deps = spec.runtime_dependencies.collect(&:name)
+        deps.intersection(gems_to_remove).each do |missing_dependency|
+          memo[missing_dependency] ||= []
+          memo[missing_dependency] << spec.name
         end
       end
-    end
+      if unsatisfied_dependency_mapping.any?
+        unsatisfied_dependency_mapping.each { |gem_to_remove, gems_depending_on_removed| display_cant_remove_message(gem_to_remove, gems_depending_on_removed) }
+        LogStash::PluginManager.ui.info("No plugins were removed.")
+        return false
+      end
```

Comment:
I gave up on the pre-validation "optimistic removal" approach.

Bundler did _not_ like being told to materialize a dependency graph that had missing dependencies (e.g., the gems we had optimistically removed), which prevented us from finding the conflicts to present them in a way consistent with how we have previously done.

Instead, this approach is a lot closer to the original. It looks at the existing dependency graph to identify which dependencies would be unmet _before_ removing them from the gemfile.